internals/secrethub/audit.go

```diff
@@ -34,27 +43,182 @@ func (cmd *AuditCommand) Register(r command.Registerer) {
 
 // Run prints all audit events for the given repository or secret.
 func (cmd *AuditCommand) Run() error {
+	cmd.beforeRun()
+	return cmd.run()
+}
+
+// beforeRun configures the command using the flag values.
+func (cmd *AuditCommand) beforeRun() {
+	cmd.timeFormatter = NewTimeFormatter(cmd.useTimestamps)
+}
+
+// Run prints all audit events for the given repository or secret.
+func (cmd *AuditCommand) run() error {
+	var auditTable auditTable
+	var iter secrethub.AuditEventIterator
+
 	repoPath, err := cmd.path.ToRepoPath()
 	if err == nil {
-		auditRepoCommand := AuditRepoCommand{
-			io:            cmd.io,
-			path:          repoPath,
-			useTimestamps: cmd.useTimestamps,
-			newClient:     cmd.newClient,
+		client, err := cmd.newClient()
+		if err != nil {
+			return err
+		}
+		tree, err := client.Dirs().GetTree(repoPath.GetDirPath().Value(), -1, false)
+		if err != nil {
+			return err
+		}
+
+		iter = client.Repos().EventIterator(repoPath.Value(), nil)
+		auditTable = newRepoAuditTable(tree, cmd.timeFormatter)
+	} else {
+		secretPath, err := cmd.path.ToSecretPath()
+		if err == nil {
+			if cmd.path.HasVersion() {
+				return ErrCannotAuditSecretVersion
+			}
+
+			client, err := cmd.newClient()
+			if err != nil {
+				return err
+			}
+
+			iter = client.Secrets().EventIterator(secretPath.Value(), nil)
+			auditTable = newSecretAuditTable(cmd.timeFormatter)
+		} else {
+			return ErrNoValidRepoOrSecretPath
 		}
-		return auditRepoCommand.Run()
 	}
 
-	secretPath, err := cmd.path.ToSecretPath()
-	if err == nil {
-		auditSecretCommand := AuditSecretCommand{
-			io:            cmd.io,
-			path:          secretPath,
-			useTimestamps: cmd.useTimestamps,
-			newClient:     cmd.newClient,
+	tabWriter := tabwriter.NewWriter(cmd.io.Stdout(), 0, 4, 4, ' ', 0)
+	header := strings.Join(auditTable.header(), "\t") + "\n"
+	fmt.Fprint(tabWriter, header)
+
+	// interactive mode is assumed, except when output is piped.
+	interactive := !cmd.io.Stdout().IsPiped()
+
+	i := 0
+	for {
+		i++
+		event, err := iter.Next()
+		if err == iterator.Done {
+			break
+		} else if err == api.ErrSecretNotFound {
+			// Check if we're attempting to audit a dir.
+			client, err := cmd.newClient()
+			if err != nil {
+				return err
+			}
+
+			_, err = client.Dirs().GetTree(api.DirPath(cmd.path).Value(), 1, false)
+			if err == nil {
+				return ErrCannotAuditDir
+			}
+			return api.ErrSecretNotFound
```

Comment:
https://github.com/secrethub/secrethub-go/pull/149 helps a bit in readability as it makes the Exists check more explicit.

For more separation, we could check for dir existence on line 90, before creating the iterator.
However, this would mean an extra HTTP call always, instead of only doing it when the secret is not found.

Reply:
We've decided to do both.